# Introduction to ChIP-seq analysis

Welcome in this tutorial, you will find here a detailed ChIP-seq workflow, starting from sequencing read to the final coverage tracks and differentially accessible genomic regions.

This tutorial is using the Galaxy platform to perform the data download, quality control, mapping and peak calling. We will then explore the result via IGV and RSAT.

## 1. A brief note on ChIP-seq

ChIP-seq stands for **Ch**romatin **I**mmuno **P**recipitation followed by **seq**uencing.

## 2. A brief note on Galaxy

Today, we will work on the Galaxy platform. It's simple, free and open-source.

## 3. Let's start the analysis : downloading the data

We will work on the study from X et al.

<details>
  <summary>Find the SRA accession identifiers</summary>

  You are looking for a code starting with `GSE`. You usually find it in the *Data accessibility* section of an article.

  </details>


## 4. Quality Checking

We need to assess the quality of the sequencing reads.
